pkg/metrics/collector.go

```diff
@@ -0,0 +1,111 @@
+package metrics
+
+import (
+	"sync"
+
+	"k8s.io/component-base/metrics"
+)
+
+type Collector struct {
+	metrics.BaseStableCollector
+	lock sync.RWMutex
+
+	storedMetrics    []metrics.Metric
+	staleMetrics     []metrics.Metric
+	markStaleMetrics bool
+}
+
+var _ metrics.StableCollector = &Collector{}
+
+const (
+	VersionLabel    = "version"
+	BuildLabel      = "build"
+	ApiVersionLabel = "api_version"
+	vCenterUUID     = "uuid"
+
+	HwVersionLabel   = "hw_version"
+	cbtMismatchLabel = "cbt"
+)
+
+// Metrics in this package can be used when we expect metrics with certain label to stop
+// being reported when cluster configuration changes.
+var (
+	EsxiVersionMetric = metrics.NewDesc(
+		"vsphere_esxi_version_total",
+		"Number of ESXi hosts with given version.",
+		[]string{VersionLabel, ApiVersionLabel}, nil,
+		metrics.ALPHA, "",
+	)
+	HwVersionMetric = metrics.NewDesc(
+		"vsphere_node_hw_version_total",
+		"Number of vSphere nodes with given HW version.",
+		[]string{HwVersionLabel}, nil,
+		metrics.ALPHA, "",
+	)
+
+	CbtMismatchMetric = metrics.NewDesc(
+		"vsphere_vm_cbt_checks",
+		"Boolean metric based on whether ctkEnabled is consistent or not across all nodes in the cluster.",
+		[]string{cbtMismatchLabel}, nil,
+		metrics.ALPHA, "",
+	)
+)
+
+func NewMetricsCollector() *Collector {
+	return &Collector{
+		storedMetrics:    []metrics.Metric{},
+		staleMetrics:     []metrics.Metric{},
+		markStaleMetrics: false,
+	}
+}
+
+func (c *Collector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	ch <- EsxiVersionMetric
+	ch <- HwVersionMetric
+	ch <- CbtMismatchMetric
+}
+
+func (c *Collector) CollectWithStability(ch chan<- metrics.Metric) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	if c.markStaleMetrics {
+		for _, m := range c.staleMetrics {
+			ch <- m
+		}
+	} else {
+		for _, m := range c.storedMetrics {
+			ch <- m
+		}
+	}
+}
```

Comment:
I like the idea of keeping one list of finished metrics from the last check and a separate list that is being collected right now. However, I don't like the naming and `markStaleMetrics`.

What if there was one complete list of metrics, say `currentMetrics` and that list would be used to feed Prometheus collector. There is nothing stale in this list, it's the latest _complete_ set.

And then say `inProgressMetrics` that is being populated right now and may not be complete. In `FinishedAllChecks` you cant put `inProgressMetrics` to `currentMetrics` and clear `inProgressMetrics` for  the next round, nothing will read from it.

Reply:
Should work one downside is - if we could not finish the checks for one error or the other, we might have to report whatever we collected so far. Not a bad tradeoff given that, currently we will be reporting stale metrics until checks can finish actually.